File: lista1/1zad.py
import sys

def search(T : str, P : str) -> list:
    m = 0
    i = 0
    results = []
    lps = find_lps(P)
    while m + i < len(T):
        if P[i] == T[m+i]:
            i += 1
            if i == len(P):
                results.append(m)
                m += 1
                # m = m + i
                i = 0
        else:
            m = m + i - lps[i]
            if i > 0:
                i = lps[i]
    for element in results:
        print("Znaleziono dopasowanie na indeksie ", element)
    return results

def find_lps(P : str) -> list:
    lps = []
    lps.append(-1)
    lps.append(0)
    i = 2
    j = 0
    while i < len(P):
        if P[i-1] == P[j]:
            lps.append(j+1)
            i += 1
            j += 1
        elif j > 0:
            j = lps[j]
        else:
            lps.append(0)
            i += 1
    return lps

# pattern = "asd"
# txt = "khbjasdjhbbjhbm sadasdasdasdasdmn asd"
# results = search(txt, pattern)

results = search(sys.argv[2], sys.argv[1])

File: lista1/1zad.cpp
#include <iostream>
#include <cstdlib>
#include <cstring>
using namespace std;

void find_lps(char* P, int* lps) {
    int P_len = strlen(P);
    lps[0] = -1;
    lps[1] = 0;
    int i = 2;
    int j = 0;

    while(i < P_len) {
        if(P[i-1] == P[j]) {
            lps[i] = j+1;
            i++;
            j++;
        }
        else if(j > 0) {
            j = lps[j];
        }
        else {
            lps[i] = 0;
            i++;
        }
    }
}

int* search(char* T, char* P) {
    int m = 0;
    int i = 0;
    int results = 0;
    int* results_array = (int*)calloc(1, sizeof(int));
    int T_len = strlen(T);
    int P_len = strlen(P);
    int* lps = (int*)calloc(P_len + 1, sizeof(int));

    find_lps(P, lps);

    while(m + i < T_len) {
        if(P[i] == T[m + i]) {
            i++;
            if(i == P_len) {
                results++;
                results_array = (int*)realloc(results_array, results * sizeof(int));
                results_array[results - 1] = m;
                m = m + i;
                i = 0;
            }
        }
        else {
            m = m + i - lps[i];
            if(i > 0) {
                i = lps[i];
            }
        }
    }
    for(int j = 0; j < results; j++) {
        cout << "Znaleziono dopasowanie na indeksie " << results_array[j] <<endl;
        // printf("Znaleziono dopasowanie na indeksie %d\n", results_array[j]);
    }
    return results_array;
}

int main(int argc, char**  argv) {
    char* P = argv[1];
    char* T = argv[2];
    search(T, P);
}